docs: add local deployment docs for en (#19)
cloudberrydb/cloudberrydb-site: docs/cbdb-op-deploy-guide.md

    #### will be created.
    COORDINATOR_DIRECTORY=/data0/master
    
    #### Port number for the coordinator instance. 
    COORDINATOR_PORT=5432
    
    #### Shell utility used to connect to remote hosts.
    TRUSTED_SHELL=ssh
    
    #### Default server-side character set encoding.
    ENCODING=UNICODE
    ```

    If there are segment mirror nodes, you also need to modify `MIRROR_PORT_BASE` and `MIRROR_DATA_DIRECTORY`.

    - `MIRROR_PORT_BASE` is the port used by the mirror.
    - `MIRROR_DATA_DIRECTORY` is the data directory of the mirror, that is, `data0/mirror` in Step 2 of [Step 4: Create data directory](#minimal-step-4).

    The modified example configuration file is as follows:

    ```toml
    #### Base number by which mirror segment port numbers 
    #### are calculated.
    MIRROR_PORT_BASE=17000
    
    #### File system location(s) where mirror segment data directories 
    #### will be created. The number of mirror locations must equal the
    #### number of primary locations as specified in the 
    #### DATA_DIRECTORY parameter.
    declare -a MIRROR_DATA_DIRECTORY=(data0/mirror)
    ```

#### Step 6: Initialize the database

> Initialize the database using the `gpinitsystem` command:

Reviewer: ```suggestion
Initialize the database with the `gpinitsystem` command:
```

```bash
gpinitsystem -c gpinitsystem_config -p cbdb_etcd.conf -h seg_host
```

### Single-node deployment mode

Single-node deployment mode means deploying FTS and ETCD services on a single local node. This mode is mainly used for code development in test scenarios. It does not support high availability, and is not suitable for production environments.

> The single-node deployment mode is non-distributed. All services are deployed on the same physical machine. To learn the configuration required for this mode, see [development and test environment configuration](./cbdb-op-software-hardware.md#development-and-test-environment).

Author: ```suggestion
The single-node deployment mode is non-distributed. All services are deployed on the same physical machine. To learn the configuration required for this mode, see [development and test environment configuration](./cbdb-op-software-hardware.md#for-development-and-test-environments).
```

<!-- 要进行单节点部署，执行以下步骤：

1. 下载 Cloudberry Database 源代码到本地目录，下载地址为 <https://github.com/cloudberrydb/cloudberrydb>。
2. 参考 Cloudberry Database 代码仓库中 README.md 文档，编译 Cloudberry Database 源代码并安装。
3. 执行 `make create-demo-cluster` 命令。-->

:::info

Currently, the steps for this deployment mode are on the way. Please keep watching.

:::

## Production deployments

### Standard distributed deployment mode

In this deployment mode, FTS and ETCD cluster nodes are deployed on separate physical machines. The system reliability is high, supporting automatic failover of master/standby nodes, which is suitable for production environments that require high availability.

> To learn the physical configuration required for this mode, see [configuration for production environment](./cbdb-op-software-hardware.md#production-environment).

Author: ```suggestion
To learn the physical configuration required for this mode, see [configuration for production environment](./cbdb-op-software-hardware.md#for-production-environments).
```

To deploy Cloudberry Database in this mode, take the following steps.

#### Step 1: Plan the deployment

It is recommended to plan the nodes to be deployed according to the following table:

| Component       | Recommended number of physical machines | Description |
| :-------------  | :---------------- | :---------- |
| Master node     | 1                |             |

> | Standby node    | 1                | Standby nodes are used for hot backup of master nodes. |

Reviewer: ```suggestion
| Standby node    | 1                | The standby node is used for hot backup of the master node. |
```
语境中standby和master都是1，就没用复数形式。

| Segment (computing node) |  3      | It is recommended to deploy the same number of mirror nodes as computing nodes for high availability. |

> | FTS node        | 3                | The FTS cluster supports multi-node independent deployment, with 3 nodes as the default configuration to ensure high availability. |

Reviewer: ```suggestion
| FTS node        | 3                | The FTS cluster supports multi-node independent deployment, with the default configuration of 3 nodes to ensure high availability. |
```

> | ETCD metadata node | 3             | The ETCD cluster supports multi-node independent deployment, with high availability features natively supported by the application. |

Reviewer: ```suggestion
| ETCD (metadata node) | 3             | The ETCD cluster supports multi-node independent deployment. High availability is natively supported by applications. |
```

:::info

> - The following deployment operations are all performed with the `gpadmin` user.

Reviewer: ```suggestion
- The following deployment operations are performed by the `gpadmin` user.
```

> - It is recommended to deploy the FTS nodes and ETCD nodes on the same server to save resources.

Reviewer: ```suggestion
- It is recommended to deploy FTS nodes and ETCD nodes on the same server to save resources.
```

:::

#### Step 2: Prepare configuration files

> 1. Prepare the corresponding physical machines according to the number of nodes planned in the previous step. Edit the `/etc/hosts` file. Then add all cluster nodes including all IP addresses and aliases. For example:

Reviewer: ```suggestion
1. Prepare physical machines according to the recommended nodes in the previous step. Edit the `/etc/hosts` file. Then add all cluster nodes including all IP addresses and aliases. For example:
```

    ```
    # Loopback address       Hostname
    <ip_address 1>           master
    <ip_address 2>           standby 1
    <ip_address 3>           segment 1
    <ip_address 4>           segment 2
    <ip_address 5>           segment 3
    <ip_address 6>           etcd 1
    <ip_address 7>           etcd 2
    <ip_address 8>           etcd 3
    ```

2. On the master node, create a file named `all_host`, and fill in all the hostnames in this file.

    ```
    master
    standby
    segment 1
    segment 2
    segment 3
    etcd 1
    etcd 2
    etcd 3
    ```

3. On the master node, create a file named `seg_host`, and fill in all the hostnames of the segment nodes in this file.

    ```
    segment 1
    segment 2
    segment 3
    ```

4. On the master node, create a file named `fts_service.conf`, and fill in all the hostnames of the FTS nodes.

    :::note

>     To save hardware resources, you do not need to deploy the FTS service in a separate cluster. It is recommended to co-deploy it with the ETCD service on the ETCD servers.

Reviewer: ```suggestion
    To save hardware resources, the FTS service does not need to be deployed in a separate cluster. It is recommended to co-deploy it with the ETCD service on the ETCD servers.
```

    :::

    ```
    etcd 1
    etcd 2
    etcd 3
    ```

5. On the master node, create a file named `etcd_service.conf`, and fill in all the hostnames of the ETCD nodes.

    ```
    etcd 1
    etcd 2
    etcd 3
    ```

6. On the master node, create a file named `cbdb_etcd.conf`. Then write the following configuration items in the file.

>     - `gp_etcd_endpoints`: Node names of the ETCD cluster service. You need to configure the ETCD service node hosts `{etcd-service-0}`, `{etcd-service-1}`, and `{etcd-service-2}` in `etcd_service.conf`. The ETCD service will start on these 3 hosts by default.

Reviewer: ```suggestion
    - `gp_etcd_endpoints`: Node names of the ETCD service. You need to configure the ETCD service node hosts `{etcd-service-0}`, `{etcd-service-1}`, and `{etcd-service-2}` in `etcd_service.conf`. The ETCD service will start on these 3 hosts by default.
```

    - `gp_etcd_account_id`: Tenant ID. You can generate a globally unique UUID using the UUID tool for it.
    - `gp_etcd_cluster_id`: Cluster ID. You can generate a globally unique UUID using the UUID tool for it.

    An example of the configuration file is as follows:

    ```conf
    gp_etcd_endpoints='{etcd-service-0}:2379,{etcd-service-1}:2379,{etcd-service-2}:2379'
    gp_etcd_account_id='e3cb5400-9589-918d-c178-82d500deac6e'
    gp_etcd_cluster_id='7bc05356-67f9-49fe-804e-12fe30b093ef'
    gp_etcd_namespace='default'
    ```

    :::info

>     `gp_etcd_namespace` is the namespace configuration for the cluster, and on-premises deployments use the default configuration.

Reviewer: ```suggestion
    `gp_etcd_namespace` is the namespace configuration of the cluster, and its default configuration can be used for on-premises deployments.
```

    ```bash
    # For example, /data0/primary and /data0/mirror.

    gpssh -f seg_host -e 'mkdir -p /data0/primary'
    gpssh -f seg_host -e 'mkdir -p /data0/mirror'
    ```

3. Create a data directory on the master node.

    ```bash
    # For example, /data0/master.

    mkdir -p /data0/master
    ```

4. Create a data directory on the standby node.

    ```bash
    # For example, /data0/master.

    gpssh -h standby -e 'mkdir -p /data0/master'
    ```

5. In the `~/.bashrc` file of master and standby nodes, add the following line of command, which is the path `{path in the previous step}` + `gpseg-1`.

    ```bash
    export COORDINATOR_DATA_DIRECTORY=/data0/master/gpseg-1
    ```

6. Run the following command to make `COORDINATOR_DATA_DIRECTORY` effective.

    ```bash
    source ~/.bashrc
    ```

#### Step 5: Configure the `gpinitsystem_config` startup script

1. On the master node, copy the template configuration file to the current directory.

    ```bash
    cp $GPHOME/docs/cli_help/gpconfigs/gpinitsystem_config .
    ```

2. Edit the `gpinitsystem_config` file:

>     - Modify the `DATA_DIRECTORY` configuration item to the data directory of the segment node, that is, `/data0/primary` in Step 2 of [Step 4: Create data directory](#standard-step-4).

Reviewer: ```suggestion
    - Modify `DATA_DIRECTORY` to the data directory of the segment node, which is `/data0/primary` as mentioned in step 2 of [Step 4: Create data directory](#standard-step-4).
```

>     - Modify `COORDINATOR_HOSTNAME` to the hostname of the master primary node.

Reviewer: ```suggestion
    - Modify `COORDINATOR_HOSTNAME` to the hostname of the master node.
```

>     - Modify `COORDINATOR_DIRECTORY` to the data directory of the master primary node, that is, `/data0/master` in Step 3 of [Step 4: Create data directory](#standard-step-4).

Reviewer: ```suggestion
    - Modify `COORDINATOR_DIRECTORY` to the data directory of the master node, which is `/data0/master` as mentioned in step 3 of [Step 4: Create data directory](#standard-step-4).
```

    The modified example configuration file is as follows:

    ```toml
    #### Base number by which primary segment port numbers 
    #### are calculated.
    PORT_BASE=6000
    
    #### File system location(s) where primary segment data directories 
    #### will be created. The number of locations in the list dictate
    #### the number of primary segments that will get created per
    #### physical host (if multiple addresses for a host are listed in 
    #### the hostfile, the number of segments will be spread evenly across
    #### the specified interface addresses).
    declare -a DATA_DIRECTORY=(/data0/primary)
    
    #### OS-configured hostname or IP address of the coordinator host.
    COORDINATOR_HOSTNAME=master
    
    #### File system location where the coordinator data directory 
    #### will be created.
    COORDINATOR_DIRECTORY=/data0/master
    
    #### Port number for the coordinator instance. 
    COORDINATOR_PORT=5432
    
    #### Shell utility used to connect to remote hosts.
    TRUSTED_SHELL=ssh
    
    #### Default server-side character set encoding.
    ENCODING=UNICODE
    ```

>     If there are segment mirror nodes, you also need to modify `MIRROR_PORT_BASE` and `MIRROR_DATA_DIRECTORY`.

Reviewer: ```suggestion
    If there are segment and mirror nodes, `MIRROR_PORT_BASE` and `MIRROR_DATA_DIRECTORY` also need to be modified.
```

>     - `MIRROR_PORT_BASE` is the port used by the mirror.

Reviewer: ```suggestion
    - `MIRROR_PORT_BASE` is the port used by the mirror node.
```

>     - `MIRROR_DATA_DIRECTORY` is the data directory of the mirror, that is, `data0/mirror` in Step 2 of [Step 4: Create data directory](#standard-step-4).

Reviewer: ```suggestion
    - `MIRROR_DATA_DIRECTORY` is the data directory of the mirror node, which is `/data0/mirror` as mentioned in step 2 of [Step 4: Create data directory](#standard-step-4).
```

    The modified example configuration file is as follows:

    ```toml
    #### Base number by which mirror segment port numbers 
    #### are calculated.
    MIRROR_PORT_BASE=17000
    
    #### File system location(s) where mirror segment data directories 
    #### will be created. The number of mirror locations must equal the
    #### number of primary locations as specified in the 
    #### DATA_DIRECTORY parameter.
    declare -a MIRROR_DATA_DIRECTORY=(data0/mirror)
    ```

#### Step 6: Initialize the database

> Initialize the database using the `gpinitsystem` command:

Reviewer: ```suggestion
Initialize the database with the `gpinitsystem` command:
```

```bash
gpinitsystem -c gpinitsystem_config -p cbdb_etcd.conf -F fts_service.conf  -E etcd_service.conf -h seg_host -s standby
```

### Hybrid deployment mode

> In this mode, you can reuse the existing physical machines to deploy the FTS and ETCD services. No additional physical machines are required for FTS and ETCD. The system reliability is not as high as that of the standard distributed deployment. This deployment mode supports automatic fault recovery by switching between master/standby nodes.

Author: 这句话本身是没意义的，我在中文版里也删掉这句话了。

> To learn the physical configuration required for this mode, see [configuration for production environment](./cbdb-op-software-hardware.md#production-environment).

Author: ```suggestion
To learn the physical configuration required for this mode, see [configuration for production environment](./cbdb-op-software-hardware.md#for-production-environments).
```

It is recommended to plan the nodes to be deployed according to the following table:

| Component       | Recommended number of physical machines | Description |
| :-------------  | :---------------- | :---------- |
| Master node     | 1                |             |
| Standby node    | 1                | Standby nodes are used for hot backup of master nodes. |
| Segment (computing node) |  3      | It is recommended to deploy the same number of mirror nodes as computing nodes for high availability. The number of nodes is determined according to your data requirements and related hardware (the number of disks mounted on the machine). |
| FTS node        | 3                | The FTS cluster supports multi-node hybrid deployment, with 3 nodes as the default configuration to ensure high availability. You do not need to reserve a separate physical machine for it. |
| ETCD metadata node | 3             | The ETCD cluster supports multi-node deployment, with high availability feature natively supported by the application. You do not need to reserve a separate physical machine for it. |

:::info

- The following deployment operations are all performed with the `gpadmin` user.

> - Because segments are computing nodes, deploying the ETCD and FTS services on computing nodes might cause performance issues in production environments. To avoid performance issues and improve system reliability, it is recommended to specify an additional physical machine `{host}` through the configuration file to deploy ETCD and FTS services, that is, master + standby + {host} deployment.

Author: 在 [Google developer documentation style guide](https://developers.google.com/style/word-list#letter-b) 中，"since" 是有歧义的单词，也可以表示时间关系。
![image](https://github.com/cloudberrydb/cloudberrydb-site/assets/41534398/18134966-e191-456d-a7ca-26cbe1617db8)

    ```
    master
    standby
    segment 1
    segment 2
    segment 3
    ```

3. On the master node, create a file named `seg_host`, and fill in all the hostnames of the segment nodes in this file.

    ```
    segment 1
    segment 2
    segment 3
    ```

4. On the master node, create a file named `cbdb_etcd.conf`. Then write the following configuration items in the file.

    - `gp_etcd_endpoints`: Node names of the ETCD cluster service. You need to configure the node names with the hostnames of master, standby, and the segment on which ETCD are to be deployed. The ETCD service will start on these 3 hosts by default.
    - `gp_etcd_account_id`: Tenant ID. You can generate a globally unique UUID using the UUID tool for it.
    - `gp_etcd_cluster_id`: Cluster ID. You can generate a globally unique UUID using the UUID tool for it.

    An example of the configuration file is as follows:

    ```conf
    gp_etcd_endpoints='master:2379,standby:2379,seg1:2379'
    gp_etcd_account_id='e3cb5400-9589-918d-c178-82d500deac6e'
    gp_etcd_cluster_id='7bc05356-67f9-49fe-804e-12fe30b093ef'
    gp_etcd_namespace='default'
    ```

    :::info

>     `gp_etcd_namespace` is the namespace configuration for the cluster, and on-premises deployments use the default configuration.

Reviewer: ```suggestion
    `gp_etcd_namespace` is the namespace configuration for the cluster. For on-premises deployments, use the default configuration of `gp_etcd_namespace`.
```

    :::

> #### Step 3: Configure SSH keyless authentication for the `gpadmin` account

Reviewer: ```suggestion
#### Step 3: Configure SSH password-free login for the `gpadmin` account
```

    ```bash
    # For example, /data0/primary and /data0/mirror.

    gpssh -f seg_host -e 'mkdir -p /data0/primary'
    gpssh -f seg_host -e 'mkdir -p /data0/mirror'
    ```

3. Create a data directory on the master node.

    ```bash
    # For example, /data0/master.

    mkdir -p /data0/master
    ```

4. Create a data directory on the standby node.

    ```bash
    # For example, /data0/master.

    gpssh -h standby -e 'mkdir -p /data0/master'
    ```

5. In the `~/.bashrc` file of master and standby nodes, add the following line of command, which is the path `{path in the previous step}` + `gpseg-1`.

    ```bash
    export COORDINATOR_DATA_DIRECTORY=/data0/master/gpseg-1
    ```

6. Run the following command to make `COORDINATOR_DATA_DIRECTORY` effective.

    ```bash
    source ~/.bashrc
    ```

#### Step 5: Configure the `gpinitsystem_config` startup script

1. On the master node, copy the template configuration file to the current directory.

    ```bash
    cp $GPHOME/docs/cli_help/gpconfigs/gpinitsystem_config .
    ```

2. Edit the `gpinitsystem_config` file:

>     - Modify the `DATA_DIRECTORY` configuration item to the data directory of the segment node, that is, `/data0/primary` in Step 2 of [Step 4: Create data directory](#hybrid-step-4).

Reviewer: ```suggestion
    - Modify `DATA_DIRECTORY` to the data directory of the segment node, which is `/data0/primary` as mentioned in step 2 of [Step 4: Create data directory](#hybrid-step-4).
```

>     - Modify `COORDINATOR_HOSTNAME` to the hostname of the master primary node.

Reviewer: ```suggestion
    - Modify `COORDINATOR_HOSTNAME` to the hostname of the master node.
```

>     - Modify `COORDINATOR_DIRECTORY` to the data directory of the master primary node, that is, `/data0/master` in Step 3 of [Step 4: Create data directory](#hybrid-step-4).

Reviewer: ```suggestion
    - Modify `COORDINATOR_DIRECTORY` to the data directory of the master node, which is `/data0/master` as mentioned in step 3 of [Step 4: Create data directory](#hybrid-step-4).
```

    The modified example configuration file is as follows:

    ```toml
    #### Base number by which primary segment port numbers 
    #### are calculated.
    PORT_BASE=6000

    #### File system location(s) where primary segment data directories 
    #### will be created. The number of locations in the list dictate
    #### the number of primary segments that will get created per
    #### physical host (if multiple addresses for a host are listed in 
    #### the hostfile, the number of segments will be spread evenly across
    #### the specified interface addresses).
    declare -a DATA_DIRECTORY=(/data0/primary)

    #### OS-configured hostname or IP address of the coordinator host.
    COORDINATOR_HOSTNAME=master

    #### File system location where the coordinator data directory 
    #### will be created.
    COORDINATOR_DIRECTORY=/data0/master

    #### Port number for the coordinator instance. 
    COORDINATOR_PORT=5432

    #### Shell utility used to connect to remote hosts.
    TRUSTED_SHELL=ssh

    #### Default server-side character set encoding.
    ENCODING=UNICODE
    ```

>     If there are segment mirror nodes, you also need to modify `MIRROR_PORT_BASE` and `MIRROR_DATA_DIRECTORY`.

Reviewer: ```suggestion
    If there are mirror nodes, `MIRROR_PORT_BASE` and `MIRROR_DATA_DIRECTORY` also need to be modified.
```

>     - `MIRROR_PORT_BASE` is the port used by the mirror.

Reviewer: ```suggestion
    - `MIRROR_PORT_BASE` is the port used by the mirror node.
```

>     - `MIRROR_DATA_DIRECTORY` is the data directory of the mirror, that is, `data0/mirror` in Step 2 of [Step 4: Create data directory](#hybrid-step-4).

Reviewer: ```suggestion
    - `MIRROR_DATA_DIRECTORY` is the data directory of the mirror node, which is `/data0/mirror` as mentioned in step 2 of [Step 4: Create data directory](#hybrid-step-4).
```

    The modified example configuration file is as follows:

    ```toml
    #### Base number by which mirror segment port numbers 
    #### are calculated.
    MIRROR_PORT_BASE=17000
    
    #### File system location(s) where mirror segment data directories 
    #### will be created. The number of mirror locations must equal the
    #### number of primary locations as specified in the 
    #### DATA_DIRECTORY parameter.
    declare -a MIRROR_DATA_DIRECTORY=(data0/mirror)
    ```

#### Step 6: Initialize the database

> Initialize the database using the `gpinitsystem` command:

Reviewer: ```suggestion
Initialize the database with the `gpinitsystem` command:
```

```bash
gpinitsystem -c gpinitsystem_config -p cbdb_etcd.conf -h seg_host -s standby
```

## Points to note

> - If you want to enable the feature that automatically detects whether the master is down and upgrades the standby to master, you need to execute the DML twice in the system.

Reviewer: ```suggestion
- If you want to enable the feature that automatically detects whether the master is valid and upgrades the standby to master, execute the DML command twice.
```

> - The ETCD service needs at least 2 live nodes to operate normally, while the FTS service only needs 1 live node to work correctly.

Reviewer: ```suggestion
- The ETCD service requires at least 2 live nodes to operate normally, while the FTS service only requires 1 live node.
```
